Enable user case creation for Web Users

dominicochoa-dimagi opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
These changes enable the creation of user cases for web users. When a web user is saved after signup, a user case is then created with the relevant info. 

~~As this PR stands currently, the changes result in user cases being saved after user creation. However, the user case~~
~~does not appear in the case list. This is true for both "web-users" and "commcare-users."~~

~~Before we merge this PR the above issue needs to be resolved.~~

This was just a local ES issue and was fixed with `./manage.py ptop_reindexer_v2 sql-case`

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/USH-943?atlOrigin=eyJpIjoiNDhmNmU2NzE4OWQzNDE5Nzk3ZGNiMjBmMTJkNjc0MWUiLCJwIjoiaiJ9)

Previously, the `sync_usercases` code was dependent on the `CommCareUser` model object. There are several feature flag checks throughout the code that call `CommCareUser.project` to retrieve the relevant domain. 

Because `WebUser` objects do not have a `project` property containing the domain name, the first issue was retrieving the relevant domain(s). `sync_webuser_usercases_if_applicable` was created which uses the `WebUser.get_domains` property to iterate over all of the domains the web user has membership. 

```python
def sync_webuser_usercases_if_applicable(user_id, spawn_task):
    user = CouchUser.get_by_user_id(user_id)
    for domain in user.get_domains():
        domain_obj = Domain.get_by_name(domain)
        if domain_obj.call_center_config.enabled or domain_obj.usercase_enabled:
            if spawn_task:
                sync_webuser_usercases_task.delay(user, domain_obj)
            else:
                sync_webuser_usercases_task(user, domain_obj)
```

The next issue was addressing the need to change the downstream feature flag checks to use a separate domain object rather than calling `CommCareUser.project`. A bulk of the changes for this ticket revolve around changing some of the helper functions to now have a `domain_obj` argument which is then used for the needed checks.

Thus: 

```python
def _iter_sync_usercase_helpers(user):
    if user.project.usercase_enabled:
        yield _get_sync_usercase_helper(
            user,
            USERCASE_TYPE,
            user.get_id
        )
```

becomes:

```python
def _iter_sync_usercase_helpers(user, domain_obj):
    usercase_type = WEB_USER_USERCASE_TYPE if user.is_web_user() else USERCASE_TYPE 
    if domain_obj.usercase_enabled:
        yield _get_sync_usercase_helper(
            user,
            domain_obj,
            usercase_type,
            user.get_id
        )
```

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
usercases & USH_web_user_case_creation (new)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Works safely locally and wrote test coverage in an already existing unit test. Also updated some other unit tests that were mildly effected. Changes are limited to usercase syncing. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
https://github.com/dimagi/commcare-hq/blob/49d71dae375980a4e967d554cdb7b1142f87d3fa/corehq/form_processor/tests/test_usercase_dbaccessor.py#L11-L35

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Commit:
added web user usercase creation funcionality

Commit:
Merge branch 'do/create-usercase-web-users' of https://github.com/dimagi/commcare-hq into do/create-usercase-web-users

snopoke commented on corehq/apps/users/models.py:
```diff
@@ -1453,8 +1453,10 @@ def save(self, fire_signals=True, **params):
+            from corehq.apps.callcenter.tasks import sync_webuser_usercases_if_applicable
             results = couch_user_post_save.send_robust(sender='couch_user', couch_user=self)
             log_signal_errors(results, "Error occurred while syncing user (%s)", {'username': self.username})
+            sync_webuser_usercases_if_applicable(self._id, spawn_task=True)
```
since this is called directly you might as well pass in the user object instead of just the ID. That will save having to fetch it again.

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
+            if spawn_task:
+                sync_webuser_usercases_task.delay(user, domain_obj)
+            else:
+                sync_webuser_usercases_task(user, domain_obj)
```
Might as well just call `sync_usercases(user, domain_obj)` directly.

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
+        domain_obj = Domain.get_by_name(domain)
+        if domain_obj.call_center_config.enabled or domain_obj.usercase_enabled:
+            if spawn_task:
+                sync_webuser_usercases_task.delay(user, domain_obj)
```
Here you want to pass in the User ID and have the task re-fetch the user when it runs. This has 2 advantages:
1. the task get's a fresh copy of the user when it runs
2. we don't need to serialize the whole user object for the task

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
+            sync_usercases_task(user._id, user.project)
+
+
+@task(serializer='pickle', queue='background_queue')
```
once you're just passing in the user ID you can remove the custom serialization here:
```suggestion
@task(queue='background_queue')
```

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
+            sync_usercases_task.delay(user._id, user.project)
         else:
-            sync_usercases_task(user._id)
+            sync_usercases_task(user._id, user.project)
```
Just call the wrapped function directly
```suggestion
            sync_usercases(user, user.project)
```

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
     if (user.project.call_center_config.enabled or user.project.usercase_enabled):
         if spawn_task:
-            sync_usercases_task.delay(user._id)
+            sync_usercases_task.delay(user._id, user.project)
```
Similar to above, pass in the domain name rather than the whole domain:
```suggestion
            sync_usercases_task.delay(user._id, user.domain)
```

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
+    return None
 
 
 @task(serializer='pickle', queue='background_queue')
```
once domain name is being passed in instead of the whole domain:
```suggestion
@task(queue='background_queue')
```

snopoke reviewed (commented):
Definitely on the right track here. One thing to mention is that I don't think we should support 'call-center' features for Web Users.

orangejenny commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_webuser_usercases_if_applicable(user_id, spawn_task):
```
Nitpick: make `web_user` two words

orangejenny commented on corehq/apps/callcenter/sync_usercase.py:
```diff
@@ -93,18 +93,17 @@ def _domain_has_new_fields(domain, field_names):
-    case = case or CaseAccessors(domain).get_case_by_domain_hq_user_id(commcare_user._id, case_type)
-    close = commcare_user.to_be_deleted() or not commcare_user.is_active
-    user_case_helper = _UserCaseHelper(domain, owner_id, commcare_user._id)
+def _get_sync_usercase_helper(user, domain_obj, case_type, owner_id, case=None):
```
Agree with this, just wanted to show that it's fine to re-fetch domain objects all over the place, because we cache `Domain.get_by_name`: https://github.com/dimagi/commcare-hq/blob/fc8ca8faa838f5bbee50d164147e5b3078fbfe41/corehq/apps/domain/models.py#L607-L610

orangejenny commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
+@task(serializer='pickle', queue='background_queue')
+def sync_webuser_usercases_task(user, domain_obj):
+    sync_usercases(user, domain_obj)
+    return None
```
nitpick: don't need the explicit `return None`

Commit:
cleaned up helper functions

dominicochoa-dimagi commented:
Made the suggested changes. User cases are still not populating in 'Case List.' Any suggestions?

orangejenny commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +46,16 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, spawn_task):
```
Should this function also check `settings.UNIT_TESTING` like `sync_usercases_if_applicable` does?

This and `sync_usercases_if_applicable` are right on the borderline where it might be better to collapse them into one function. `user.get_domains()` should also work for mobile users, it'll just return their one domain.

orangejenny replied:
Looking at the test failures, checking `UNIT_TESTING` here will likely fix the one that says `Database queries to 'default' are not allowed in SimpleTestCase subclasses.` - that happens when you try to access the db in a test that doesn't allow it, which sounds like some test that didn't previously use the db is now trying to save usercases.

orangejenny commented:
You might just not have gotten here yet, but looking at the ticket, this PR is also going to need to update the `create_usercases` task to create usercases for web users when a project starts using usercase.

We'll also need to create usercases for web users in projects that are using usercase now. Once this is released, projects are going to start seeing web user usercases appear ad hoc, as web users are saved. We might want to hold off on releasing this until usercases are available in the restore. Let's chat on the ticket, so the product team can participate.

orangejenny commented:
I keep thinking of additional things. [This](https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/tests/test_usercase_dbaccessor.py) would be a good test to update, to verify that usercases are created for web users.

Commit:
fixed python errors and added feature flag

Commit:
fixed additional python errors

Commit:
fixed some python errors

Commit:
fixed some python errors

snopoke commented on corehq/apps/callcenter/sync_usercase.py:
```diff
@@ -226,32 +227,35 @@ def _call_center_location_owner(user, ancestor_level):
-            _iter_sync_usercase_helpers(user),
-            _iter_call_center_case_helpers(user),
+            _iter_sync_usercase_helpers(user, domain_obj),
+            _iter_call_center_case_helpers(user, domain_obj),
```
```suggestion
            _iter_call_center_case_helpers(user, domain_obj) if isinstance(user, CommCareUser) else [],
```

(could also push this down to `_iter_call_center_case_helpers` which might be cleaner)

dominicochoa-dimagi replied:
This is because we don't want to support call center features for Web Users correct?

snopoke replied:
yes, that's correct.

orangejenny replied:
Doesn't this mean `sync_call_center_user_case` could go back to only accepting mobile users and wouldn't need the domain passed in?

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,18 @@ def calculate_indicators():
+        return
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
+        if domain_obj.usercase_enabled and domain in USH_WEB_USER_CASE_CREATION.get_enabled_domains():
```
`get_enabled_domains` isn't cached to it should be moved outside of the `for` loop. Alternately you could use `USH_WEB_USER_CASE_CREATION.enabled(domain)`

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,18 @@ def calculate_indicators():
+    if settings.UNIT_TESTING:
+        return
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
```
I think the settings check should move into the loop:

```suggestion
        domain_obj = Domain.get_by_name(domain)
        if not domain_obj and settings.UNIT_TESTING:
            continue
```

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -58,12 +71,18 @@ def sync_usercases_if_applicable(user, spawn_task):
+@task(queue='background_queue')
+def sync_web_user_usercases_task(user_id, domain):
+    user = CouchUser.get_by_user_id(user_id)
+    sync_usercases(user, domain)
 
 
-@task(serializer='pickle', queue='background_queue')
-def sync_usercases_task(user_id):
+@task(queue='background_queue')
+def sync_usercases_task(user_id, domain):
     user = CommCareUser.get_by_user_id(user_id)
-    sync_usercases(user)
+    sync_usercases(user, domain)
```
These could be combined into a single tasks since `CouchUser.get_by_user_id` works for both web and mobile users
```python
@task(queue='background_queue')
def sync_usercases_task(user_id, domain):
    user = CouchUser.get_by_user_id(user_id)
    sync_usercases(user, domain)
```

snopoke commented on corehq/form_processor/tests/test_usercase_dbaccessor.py:
```diff
@@ -28,8 +30,15 @@ def setUp(self):
         self.assertIsNotNone(usercase)
         self.assertEqual(usercase.name, 'bar')
+
+    def test_get_web_user_usercase(self):
```
I don't think this test is necessary since the db accessor isn't aware of the user type at all and only uses the user ID and case type for the lookup.

snopoke reviewed (commented):
Still needs some tests for web users.

Commit:
removed test changes, created new unit test, cleaned sync_usercase helper functions

Commit:
created save function for WebUser model

orangejenny commented:
Recent commits look good to me. Doesn't this still need to update `create_usercases` to create cases for web users?

dominicochoa-dimagi commented:
Yes it does let me push that really quick.

dominicochoa-dimagi committed:
updated create_usercases()

dominicochoa-dimagi committed:
updated _iter_sync_usercase_helpers

Added feature flag check to `_iter_sync_usercase_helpers` in order to prevent web user usercases being created when `create_usercases` is called unless the feature flag is enabled.

orangejenny commented on corehq/apps/callcenter/sync_usercase.py:
```diff
@@ -179,27 +180,30 @@ def get_sync_lock_key(user_id):
 
 
-def _get_call_center_case_and_owner(user):
+def _get_call_center_case_and_owner(user, domain_obj):
```
Now that `_iter_call_center_case_helpers` returns early for web users, can't you remove `domain_obj` from this function and go back to using `user.domain`?

orangejenny commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
 
+def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
+    domain_obj = Domain.get_by_name(domain)
+    if not domain_obj and settings.UNIT_TESTING:
```
Why bother checking for `domain_obj` here?

dominicochoa-dimagi replied:
I was being overly safe. For some reason if `Domain.get_by_name` returned None for some inexplicable reason this would catch it. In hindsight this is probably not necessary.

orangejenny replied:
Nah, you're right, it's better to be cautious.

orangejenny commented on corehq/apps/users/models.py:
```diff
@@ -2433,6 +2433,12 @@ def get_dimagi_emails_by_domain(cls, domain):
+        super().save(fire_signals=fire_signals, **params)
+        from corehq.apps.callcenter.tasks import sync_web_user_usercases_if_applicable
+        for domain in self.get_domains():
+            sync_web_user_usercases_if_applicable(self, domain, spawn_task=True)
```
`CommCareUser.save` only calls `sync_usercases_if_applicable` if `fire_signals` is true. I'd expect this to behave the same.

orangejenny commented on corehq/toggles.py:
```diff
@@ -849,6 +849,16 @@ def _ensure_valid_randomness(randomness):
+    'web_user_case_creation',
+    "USH: Enable the creation of user cases for web users.",
+    TAG_CUSTOM,
+    help_link='.',
```
This PR looks almost ready, so it'd be a good time to create docs and add the link here. You can make a child page of [this page](https://confluence.dimagi.com/pages/viewpage.action?pageId=41485226). It doesn't need to be long - [this page](https://confluence.dimagi.com/display/saas/USH%3A+Enable+Web+User+Activity+Report) is a basic example of docs for another feature flag that's USH-owned and still under development, like your new flag.

Commit:
added documentation link, cleaned helper functions, added fire_signals check

Commit:
undoing hq-compose.yml change

Commit:
fixed sync_call_center_user_case() and related tests

dominicochoa-dimagi committed:
Update test_web_user_sync_usercases.py

dominicochoa-dimagi committed:
Update test_web_user_sync_usercases.py

snopoke commented on corehq/apps/app_manager/tasks.py:
```diff
@@ -18,8 +18,8 @@
     from corehq.apps.callcenter.sync_usercase import sync_usercase
-    for user in CommCareUser.by_domain(domain_name):
-        sync_usercase(user)
+    for user in CouchUser.by_domain(domain_name):
```
this will include web users even if the FF is not turned on:
```suggestion
    if USH_WEB_USER_CASE_CREATION.enabled(domain_name):
        users = CouchUser.by_domain(domain_name)
    else:
        users = CommCareUser.by_domain(domain_name)
    for user in users:
```

(code changes require importing the toggle)

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
```
`spawn_task` is always True so can be removed

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
 
+def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
+    domain_obj = Domain.get_by_name(domain)
+    if settings.UNIT_TESTING:
+        return
```
Somehow this just feels wrong to me. This makes this code and any calling code un-testable. I know there's not much code here to test and that the functions being called are tested elsewhere. Since this is already limited by a FF I think it would be sufficient to just check the FF:

```python
    if not USH_WEB_USER_CASE_CREATION.enabled(domain):
        return
    domain_obj = Domain.get_by_name(domain)
    if domain_obj.usercase_enabled:
        if spawn_task:
            sync_usercases_task.delay(user._id, domain)
        else:
            sync_usercases(user, domain)
```

dominicochoa-dimagi replied:
should I also change this in `sync_usercases_if_applicable`? I was just going by what is there.
```python
def sync_usercases_if_applicable(user, spawn_task):
    if settings.UNIT_TESTING and not user.project:
        return
    if (user.project.call_center_config.enabled or user.project.usercase_enabled):
        if spawn_task:
            sync_usercases_task.delay(user._id, user.domain)
        else:
            sync_usercases(user, user.domain)
```

snopoke replied:
no I don't think there's a need to change that since there is no FF enabling that feature

dominicochoa-dimagi replied:
One thing to note, the `settings.UNIT_TESTING` check was added to prevent testing errors as a result of some `SimpleTestCases` that create a web user and then throw an error when the db accessor is called.

snopoke replied:
as long as the FF is not enabled for those tests the it shouldn't be an issue

dominicochoa-dimagi replied:
I'm now realizing this was an issue before I even created the FF so I'm mistaken.

So would something like this work:
```python
def sync_web_user_usercases_if_applicable(user, domain):
    domain_obj = Domain.get_by_name(domain)
    if not domain_obj:
        return
    if domain_obj.usercase_enabled and USH_WEB_USER_CASE_CREATION.enabled(domain):
        sync_usercases_task.delay(user._id, domain)
```

snopoke replied:
I think the code should fail if the domain is `None`. The other code combines this with a check on unit testing setting so that it will fail if the condition occurs in production. My suggestion for this function is slightly different since this has a FF so we shouldn't need the settings check. Any unit test that is executing this code with the FF enabled should also be creating the domain object in the DB.

Let me know if you want to go to voice on this.

dominicochoa-dimagi replied:
Okay that makes sense. Thanks for the clarification.
I think a quick 5 mins on voice would help tie a bow on this.

orangejenny commented:
Looks good to me! Leaving approval for @snopoke once his comments are addressed.

Commit:
cleaned task function, added FF check on create_usercases()

Commit:
Merge branch 'do/create-usercase-web-users' of https://github.com/dimagi/commcare-hq into do/create-usercase-web-users

dominicochoa-dimagi committed:
Update tasks.py

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -47,15 +47,10 @@ def calculate_indicators():
     domain_obj = Domain.get_by_name(domain)
-    if settings.UNIT_TESTING:
-        return
     if domain_obj.usercase_enabled and USH_WEB_USER_CASE_CREATION.enabled(domain):
```
the FF check should come before the `domain_obj` is fetched / used. The reason being that we can avoid an unnecessary object fetch and also in tests the domain may not exist so it the variable could be `None`. For tests with the FF enabled they would always create the domain object.

snopoke commented on corehq/apps/callcenter/tasks.py:
```diff
@@ -46,6 +47,12 @@ def calculate_indicators():
 
 
+def sync_web_user_usercases_if_applicable(user, domain):
+    domain_obj = Domain.get_by_name(domain)
+    if USH_WEB_USER_CASE_CREATION.enabled(domain):
+        sync_usercases_task.delay(user._id, domain)
```
this should still check the domain config if the FF is enabled:
```suggestion
    if not USH_WEB_USER_CASE_CREATION.enabled(domain):
        return
    domain_obj = Domain.get_by_name(domain)
    if domain_obj.usercase_enabled:
        sync_usercases_task.delay(user._id, domain)
```

dominicochoa-dimagi committed:
Update corehq/apps/callcenter/tasks.py

Co-authored-by: Simon Kelly <skelly@dimagi.com>

snopoke commented on corehq/form_processor/tests/test_usercase_dbaccessor.py:
```diff
@@ -4,7 +4,7 @@
 from corehq.apps.app_manager.const import USERCASE_TYPE
 from corehq.apps.domain.models import Domain
-from corehq.apps.users.models import CommCareUser
+from corehq.apps.users.models import CommCareUser, WebUser
```
```suggestion
from corehq.apps.users.models import CommCareUser
```

dominicochoa-dimagi committed:
Update corehq/form_processor/tests/test_usercase_dbaccessor.py

Co-authored-by: Simon Kelly <skelly@dimagi.com>

esoergel commented on corehq/apps/callcenter/sync_usercase.py:
```diff
@@ -226,31 +229,36 @@ def _call_center_location_owner(user, ancestor_level):
 
-def sync_usercase(user):
+def sync_usercase(user, domain):
     with CriticalSection(get_sync_lock_key(user._id)):
```
Should this lock be user ID and domain?  I think as long as this code doesn't modify the user itself, that should be fine.  A task will be queued for each domain the user is a member of, and they'll all be fighting for the same lock.  I imagine we'd want to either let them all run in parallel (by adding domain to the lock key), or run them serially in a single task.

snopoke replied:
@dominicochoa-dimagi I think this still needs to be addressed

dominicochoa-dimagi replied:
would this work: 
```python
def get_sync_lock_key(user_id, domain):
    return [f"sync_user_case_for_{user_id}_{domain}"]
```

then:

```python
def sync_usercase(user, domain):
    with CriticalSection(get_sync_lock_key(user._id, domain)):
        domain_obj = Domain.get_by_name(domain)
        _UserCaseHelper.commit(list(_iter_sync_usercase_helpers(user, domain_obj)))
```

dominicochoa-dimagi committed:
Update toggles.py

dominicochoa-dimagi committed:
Update toggles.py

dominicochoa-dimagi committed:
Update toggles.py

Commit:
fixed references to USH_WEB_USER_CASE_CREATION

orangejenny commented:
Confirming that the last few commits, related to renaming the feature flag, look fine.

Commit:
removing commcare-cloud submodule

Commit:
removing commcare-cloud submodule again

Commit:
removing commcare-cloud submodule

snopoke commented:
@dominicochoa-dimagi what's outstanding on this PR?

dominicochoa-dimagi commented:
@snopoke this is ready. Just wanted okay for the new feature flag name. This has been tested on staging so it’s ready to merge otherwise.

Commit:
updated get_sync_lock_key and related helper functions

orangejenny commented on corehq/apps/callcenter/sync_usercase.py:
```diff
@@ -176,12 +176,12 @@ def _not_same(val1, val2):
 
-def sync_call_center_user_case(user):
-    with CriticalSection(get_sync_lock_key(user._id)):
+def sync_call_center_user_case(user, domain):
```
Not blocking: It looks like this is only used in tests. If so, it can be deleted.

orangejenny reviewed (approved):
Last commits look good. This is good to merge - I left one comment about possible dead code, but I'd suggest doing it in a followup PR.
